File: backend/livestream/domains/vod_comment.go
package domains

import (
	"context"
	response "sen1or/letslive/livestream/response"
	"time"

	"github.com/gofrs/uuid/v5"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
)

// DBTX is an interface satisfied by both *pgxpool.Pool and pgx.Tx,
// allowing repository methods to work with either.
type DBTX interface {
	Query(ctx context.Context, sql string, args ...any) (pgx.Rows, error)
	QueryRow(ctx context.Context, sql string, args ...any) pgx.Row
	Exec(ctx context.Context, sql string, args ...any) (pgconn.CommandTag, error)
}

type VODComment struct {
	Id         uuid.UUID  `json:"id" db:"id"`
	VODId      uuid.UUID  `json:"vodId" db:"vod_id"`
	UserId     uuid.UUID  `json:"userId" db:"user_id"`
	ParentId   *uuid.UUID `json:"parentId" db:"parent_id"`
	Content    string     `json:"content" db:"content"`
	IsDeleted  bool       `json:"isDeleted" db:"is_deleted"`
	LikeCount  int64      `json:"likeCount" db:"like_count"`
	ReplyCount int64      `json:"replyCount" db:"reply_count"`
	CreatedAt  time.Time  `json:"createdAt" db:"created_at"`
	UpdatedAt  time.Time  `json:"updatedAt" db:"updated_at"`
}

type VODCommentLike struct {
	Id        uuid.UUID `json:"id" db:"id"`
	CommentId uuid.UUID `json:"commentId" db:"comment_id"`
	UserId    uuid.UUID `json:"userId" db:"user_id"`
	CreatedAt time.Time `json:"createdAt" db:"created_at"`
}

type VODCommentRepository interface {
	WithTx(tx pgx.Tx) VODCommentRepository
	GetByVODId(ctx context.Context, vodId uuid.UUID, page int, limit int) ([]VODComment, *response.Response[any])
	CountByVODId(ctx context.Context, vodId uuid.UUID) (int, *response.Response[any])
	CountReplies(ctx context.Context, parentId uuid.UUID) (int, *response.Response[any])
	GetReplies(ctx context.Context, parentId uuid.UUID, page int, limit int) ([]VODComment, *response.Response[any])
	GetById(ctx context.Context, id uuid.UUID) (*VODComment, *response.Response[any])
	Create(ctx context.Context, comment VODComment) (*VODComment, *response.Response[any])
	IncrementReplyCount(ctx context.Context, commentId uuid.UUID) *response.Response[any]
	DecrementReplyCount(ctx context.Context, commentId uuid.UUID) *response.Response[any]
	SoftDelete(ctx context.Context, id uuid.UUID) *response.Response[any]
	HardDelete(ctx context.Context, id uuid.UUID) *response.Response[any]
}

type VODCommentLikeRepository interface {
	WithTx(tx pgx.Tx) VODCommentLikeRepository
	GetUserLikedCommentIds(ctx context.Context, commentIds []uuid.UUID, userId uuid.UUID) ([]uuid.UUID, *response.Response[any])
	InsertLike(ctx context.Context, commentId uuid.UUID, userId uuid.UUID) *response.Response[any]
	DeleteLike(ctx context.Context, commentId uuid.UUID, userId uuid.UUID) *response.Response[any]
	IncrementLikeCount(ctx context.Context, commentId uuid.UUID) *response.Response[any]
	DecrementLikeCount(ctx context.Context, commentId uuid.UUID) *response.Response[any]
}

File: mobile/lib/models/vod_comment.dart
class CommentUser {
  final String id;
  final String username;
  final String? displayName;
  final String? profilePicture;

  const CommentUser({
    required this.id,
    required this.username,
    this.displayName,
    this.profilePicture,
  });

  factory CommentUser.fromJson(Map<String, dynamic> json) {
    return CommentUser(
      id: json['id'] as String,
      username: json['username'] as String,
      displayName: json['displayName'] as String?,
      profilePicture: json['profilePicture'] as String?,
    );
  }
}

class VodComment {
  final String id;
  final String vodId;
  final String userId;
  final String? parentId;
  final String content;
  final bool isDeleted;
  final int likeCount;
  final int replyCount;
  final String createdAt;
  final String updatedAt;
  final CommentUser? user;

  const VodComment({
    required this.id,
    required this.vodId,
    required this.userId,
    this.parentId,
    required this.content,
    this.isDeleted = false,
    this.likeCount = 0,
    this.replyCount = 0,
    required this.createdAt,
    required this.updatedAt,
    this.user,
  });

  factory VodComment.fromJson(Map<String, dynamic> json) {
    return VodComment(
      id: json['id'] as String,
      vodId: json['vodId'] as String,
      userId: json['userId'] as String,
      parentId: json['parentId'] as String?,
      content: json['content'] as String,
      isDeleted: json['isDeleted'] as bool? ?? false,
      likeCount: json['likeCount'] as int? ?? 0,
      replyCount: json['replyCount'] as int? ?? 0,
      createdAt: json['createdAt'] as String,
      updatedAt: json['updatedAt'] as String,
      user: json['user'] != null
          ? CommentUser.fromJson(json['user'] as Map<String, dynamic>)
          : null,
    );
  }
}
